syntax = "proto3";

package disthash;

import "google/protobuf/empty.proto";

service DistHash {

    rpc Set (SetRequest) returns (google.protobuf.Empty) {}

    rpc Get (GetRequest) returns (GetResponse) {} 

    rpc RequestVote (RequestVoteArgs) returns (RequestVoteResponse) {}

    rpc AppendEntries (AppendEntriesArgs) returns (AppendEntriesResponse) {}
}

message GetRequest {
    string key = 1;
}

message GetResponse {
    string value = 1;
    bool notFound = 2;
}

message SetRequest {
    string key = 1;
    string value = 2;
}

message RequestVoteArgs {
    int32 term = 1;
    int32 candidateId = 2;
    int32 lastLogIndex = 3;
    int32 lastLogTerm = 4;
}

message RequestVoteResponse {
    int32 term = 1;
    bool voteGranted = 2;
}

message AppendEntriesArgs {
    int32 leaderterm = 1; 
    int32 leaderId = 2;
    int32 prevLogIndex = 3;
    int32 prevLogTerm = 4;
    repeated LogEntry entries = 5;
    int32 leaderCommitIndex = 6;
}

message AppendEntriesResponse {
    int32 term = 1;
    bool success = 2;
}

message LogEntry {
    SetRequest request = 1;
    int32 term = 2;
    int32 index = 3;
}

message ROS {
    // Rest of State (apart from log)
    int32 currTerm = 1;
    int32 votedFor = 2;
}

// Use python3 -m grpc_tools.protoc -I. --python_out=. --pyi_out=. --grpc_python_out=. disthash.proto